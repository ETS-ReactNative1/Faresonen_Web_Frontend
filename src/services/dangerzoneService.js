import axios from "axios";

const dangerzoneService = {
  getPlayersWithYellowCards(tournamentId) {
    if (!tournamentId) {
      throw new Error(
        "TournamentID is not defined. Please provide a tournamentId"
      );
    }
    return new Promise((resolve, reject) => {
      axios
        .get(`/statistics/yellowcards/${tournamentId}`)
        .then(data => {
          resolve({ data: data.data });
        })
        .catch(err => reject(err));
    });
  },

  sortTeams(players) {
    return players.sort((a, b) => a.name.localeCompare(b.name));
  },

  filterPlayers(players) {
    return Object.keys(players)
      .map(team => ({
        name: team,
        players: players[team].players.filter(player => {
          const yellowCards = player.value1;
          if (yellowCards === 3 || (yellowCards % 2 !== 0 && yellowCards > 2)) {
            return true;
          }
          return false;
        })
      }))
      .filter(team => team.players.length > 0);
  },

  groupPlayers(players) {
    return players.reduce((obj, elem) => {
      if (!obj[elem.team]) {
        // eslint-disable-next-line no-param-reassign
        obj[elem.team] = {
          players: []
        };
      }

      if (obj[elem.team]) {
        obj[elem.team].players.push(elem);
      }
      return obj;
    }, {});
  },

  groupPlayersArrayResponse(players) {
    return this.fromObjectToArray(this.groupPlayers(players));
  },

  fromObjectToArray(objects) {
    const list = [];
    Object.keys(objects).forEach(key => {
      list.push({ team: key, players: objects[key].players });
    });
    return list;
  }
};

export default dangerzoneService;
