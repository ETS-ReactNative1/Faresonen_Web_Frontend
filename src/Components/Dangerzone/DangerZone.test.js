import dangerzoneService from "../../services/dangerzoneService";

describe("Dangerzone", () => {
  describe("teams should be sorted alphabetically", () => {
    it("should return a list of alphabetically sorted teams", () => {
      let teams = [
        { name: "Brann" },
        { name: "Start" },
        { name: "Vålerenga" },
        { name: "Bodø/Glimt" }
      ];

      teams = dangerzoneService.sortTeams(teams);
      expect(teams).toEqual([
        { name: "Bodø/Glimt" },
        { name: "Brann" },
        { name: "Start" },
        { name: "Vålerenga" }
      ]);
    });
  });

  describe("players with even number of yellow cards should be returned", () => {
    it("should filter out players with odd number yellow cards", () => {
      const teams = {
        Brann: {
          players: [
            {
              name: "Ole",
              value1: 4
            },
            {
              name: "Petter",
              value1: -1
            },
            {
              name: "Sindre",
              value1: 0
            },
            {
              name: "Nikko",
              value1: 3
            }
          ]
        }
      };

      const filtered = dangerzoneService.filterPlayers(teams);
      expect(filtered).toEqual([
        {
          name: "Brann",
          players: [{ name: "Nikko", value1: 3 }]
        }
      ]);
    });
  });

  describe("players should be grouped by team name", () => {
    it("should group players by team name", () => {
      const players = [
        { team: "Brann", name: "Ole" },
        { team: "Rosenborg", name: "Petter" },
        { team: "Brann", name: "Sigurd" },
        { team: "Ranheim", name: "Kåre" },
        { team: "Brann", name: "Sigve" },
        { team: "Ranheim", name: "Aleks" },
        { team: "Rosenborg", name: "Steffen" },
        { team: "Brann", name: "Sindre" }
      ];

      const grouped = dangerzoneService.groupPlayers(players);
      expect(grouped).toEqual({
        Brann: {
          players: [
            { team: "Brann", name: "Ole" },
            { team: "Brann", name: "Sigurd" },
            { team: "Brann", name: "Sigve" },
            { team: "Brann", name: "Sindre" }
          ]
        },
        Ranheim: {
          players: [
            { team: "Ranheim", name: "Kåre" },
            { team: "Ranheim", name: "Aleks" }
          ]
        },
        Rosenborg: {
          players: [
            { team: "Rosenborg", name: "Petter" },
            { team: "Rosenborg", name: "Steffen" }
          ]
        }
      });
    });
  });
});
