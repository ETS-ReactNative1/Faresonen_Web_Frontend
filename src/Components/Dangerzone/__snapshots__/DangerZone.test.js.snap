// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dangerzone Should render properly 1`] = `
<div
  className="ui segment"
>
  <div
    className="ui dimmer"
    onClick={[Function]}
  >
    <div
      className="content"
    >
      <div
        className="ui text loader"
      >
        Henter spillere i faresonen
      </div>
    </div>
  </div>
  <h1>
    Faresonen i 
    Premier League
  </h1>
  <div
    className="ui four column grid"
  >
    <div
      className="column"
    >
      <div
        className="ui list similar-height"
        role="list"
      >
        <div
          className="item"
          onClick={[Function]}
          role="listitem"
        >
          <div
            className="content"
          >
            <div
              className="header"
            >
              Everton
            </div>
            <div
              className="description"
            >
              1
               spillere
            </div>
            <div
              className="list"
            >
              <div
                className="item"
                onClick={[Function]}
                role="listitem"
              >
                <i
                  aria-hidden="true"
                  className="user icon"
                  onClick={[Function]}
                />
                <div
                  className="content"
                >
                  <div
                    className="header"
                  >
                    Wayne Rooney
                  </div>
                  <div
                    className="description"
                  >
                    1
                     
                    gult kort
                    , 
                    Everton
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Dangerzone Should render properly when no players are present 1`] = `
<div
  className="ui info message no-print"
>
  <div
    className="header"
  >
    Faresonen ikke tilgjengelig
  </div>
  <p>
    Det er ingen spillere i Faresonen for
    Premier League
    .
  </p>
</div>
`;

exports[`Dangerzone Should render properly when players have more than 1 yellow card 1`] = `
<div
  className="ui segment"
>
  <div
    className="ui dimmer"
    onClick={[Function]}
  >
    <div
      className="content"
    >
      <div
        className="ui text loader"
      >
        Henter spillere i faresonen
      </div>
    </div>
  </div>
  <h1>
    Faresonen i 
    Premier League
  </h1>
  <div
    className="ui four column grid"
  >
    <div
      className="column"
    >
      <div
        className="ui list similar-height"
        role="list"
      >
        <div
          className="item"
          onClick={[Function]}
          role="listitem"
        >
          <div
            className="content"
          >
            <div
              className="header"
            >
              Everton
            </div>
            <div
              className="description"
            >
              1
               spillere
            </div>
            <div
              className="list"
            >
              <div
                className="item"
                onClick={[Function]}
                role="listitem"
              >
                <i
                  aria-hidden="true"
                  className="user icon"
                  onClick={[Function]}
                />
                <div
                  className="content"
                >
                  <div
                    className="header"
                  >
                    Wayne Rooney
                  </div>
                  <div
                    className="description"
                  >
                    3
                     
                    gule kort
                    , 
                    Everton
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
